### ISSUE TITLE
Clarification on Using Concatenated Input for R-Drop Training

### ISSUE THREAD

**Issue by xyb314** [dropreg/R-Drop#33]
Hello,

I'm currently studying your implementation of the R-Drop training algorithm and have a question regarding the use of concatenated input for Dropout.

In the original paper, it's mentioned that "we do not forward the input data twice, instead, we repeat the input data x and concatenate them ([x; x]) in the batch-size dimension, which can make forward procedure happen in the same mini-batch to save the training cost". 

I understand that despite the inputs being identical, the Dropout layer might still lead to different outputs due to the randomness in neuron dropout. However, I'm still a bit unclear on why this approach is valid when we hope to obtain two different outputs affected by Dropout differently.

Could you please provide some insights into why and how this approach works as expected? Any clarification would be greatly appreciated.

Thank you very much for your time!

Best,
xyb314

**Comment by apeterswu** [comment]
Hi,

Thanks for your interest and the question.

As you understand, in pytorch implementation of the dropout layer, each data item in a batch would get different dropped neuros for its own. Therefore, it is clear that concatenated operation will get two different dropped neuros for the same input x. 
For simpleness, if batch size is 1, the input data is x, then concatenated data is [x; x], and the resulted batch size is 2. Then one forward pass would get two different dropped neuros for the repeated x, then we can calculate kl divergence between the two resulted probabilities.

Not sure whether you understand, welcome if you have more questions.
